File: Week 01/id_011/two-sum.go
package algorithm00401

// 对撞指针解法
func twoSum2(numbers []int, target int) []int {
	l, r := 0, len(numbers)-1
	for l < r {
		if numbers[l]+numbers[r] == target {
			return []int{l, r}
		} else if numbers[l]+numbers[r] > target {
			r--
		} else {
			l++
		}
	}
	return nil
}

// hash 解法
func twoSum(numbers []int, target int) []int {
	memo := make(map[int]int)
	for i := 0; i < len(numbers); i++ {
		sub := target - nums[i]
		if j, ok := memo[sub]; ok {
			return []int{j, i}
		} else {
			memo[nums[i]] = i
		}
	}
	return nil
}

File: Week 02/id_081/TwoSum.java
/**
 * TwoSum
 */
public class TwoSum {

    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        for(int i = 0; i < nums.length; i++){
            int complement = target - nums[i];
            if(!map.containsKey(complement)) map.put(nums[i], i);
            else return new int[]{i, map.get(complement)};
        }
        return new int[]{};
    }
}